src/cunumeric/random/bitgenerator_curand.inl

```diff
@@ -0,0 +1,313 @@
+ *
+ */
+
+#pragma once
+
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <mutex>
+
+#if 0  // for debugging
+#define VERBOSE_DEBUGGING
+#endif
+
+#ifdef VERBOSE_DEBUGGING
+
+static void printtid(int op)
+{
+  pid_t tid;
+  tid = syscall(SYS_gettid);
+  ::fprintf(stderr, "[INFO-BITGENERATOR] : op = %d -- tid = %d -- pid = %d\n", op, tid, getpid());
+}
+
+#else
+
+static void printtid(int) {}
+
+#endif
+
+#include "cunumeric/random/bitgenerator.h"
+#include "cunumeric/random/bitgenerator_template.inl"
+#include "cunumeric/random/bitgenerator_util.h"
+
+#include "cunumeric/random/curand_help.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+static Logger log_curand("cunumeric.random");
+
+template <VariantKind kind>
+struct CURANDGeneratorBuilder;
+
+struct CURANDGenerator {
+  static constexpr size_t DEFAULT_DEV_BUFFER_SIZE = 64 * 1024;  // TODO: optimize this
+  curandGenerator_t gen;
+  uint64_t seed;
+  uint64_t offset;
+  curandRngType type;
+  bool supports_skipahead;
+  size_t dev_buffer_size;  // in number of entries
+  uint32_t* dev_buffer;    // buffer for intermediate results
+
+  std::mutex lock;  // in case several threads would want to use the same generator...
+
+ protected:
+  CURANDGenerator() { log_curand.debug() << "CURANDGenerator::create"; }
+
+ public:
+  virtual ~CURANDGenerator() { log_curand.debug() << "CURANDGenerator::destroy"; }
+
+  void skip_ahead(uint64_t count)
+  {
+    if (supports_skipahead) {
+      // skip ahead
+      log_curand.debug() << "[skip-ahead] : count = " << count << " - offset = " << offset;
+      offset += count;
+      CHECK_CURAND(::curandSetGeneratorOffset(gen, offset));
+    } else {
+      // actually generate numbers in the temporary buffer
+      log_curand.debug() << "[blank-generate] : count = " << count << " - offset = " << offset;
+      uint64_t remain = count;
+      while (remain > 0) {
+        if (remain < dev_buffer_size) {
+          CHECK_CURAND(::curandGenerate(gen, dev_buffer, (size_t)remain));
+          offset += remain;
+          break;
+        } else {
+          CHECK_CURAND(::curandGenerate(gen, dev_buffer, (size_t)dev_buffer_size));
+          offset += dev_buffer_size;
+        }
+        remain -= dev_buffer_size;
+      }
+    }
+  }
+
+  void generate_raw(uint64_t count, uint32_t* out)
+  {
+    CHECK_CURAND(::curandGenerate(gen, out, count));
+    offset += count;
+  }
+};
+
+struct generate_fn {
+  template <int32_t DIM>
+  size_t operator()(CURANDGenerator& gen,
+                    legate::Store& output,
+                    const DomainPoint& strides,
+                    uint64_t totalcount)
+  {
+    const auto proc = Legion::Processor::get_executing_processor();
+
+    auto rect       = output.shape<DIM>();
+    uint64_t volume = rect.volume();
+
+    for (int k = 0; k < DIM; ++k)
+      log_curand.debug() << "\t[proc=" << proc.id << "] - shape[" << k << "][" << rect.lo[k] << ":"
```

Comment:
I don't see why you marked my previous comment as resolved... Like I said, `proc` and `rect` are serializable, so unless you really mean this for loop, you can simply log them by, for example, `log_curand.debug() << proc << " " << rect`